### ISSUE TITLE
Creating derived files fails for images without file extensions

### ISSUE THREAD

**Issue by cviebrock** [spatie/laravel-medialibrary#186]
The scenario is that I've uploaded a PNG image called `blob` (note: no file extension).

The input passes Laravel's validating rule `mimes:jpg,png,gif` because, deep-down, it uses ` \Symfony\Component\HttpFoundation\File\File::getMimeType()` to figure out what kind of file it actually is, despite not having a file extension.

When it comes time for the MediaLibrary to create derived files, the first check is:

```php
if ($media->type == Media::TYPE_OTHER) {
    return;
}
```

This uses the `\Spatie\MediaLibrary\Media::getTypeAttribute` method, which only looks at the filename extension, so it skips generating derived files for that image.

It would add a bit of overhead, but would you consider changing the method to guess the mime-type of the file?  Something like this pseudo-code, which I could do by extending the Media class:

```php
class MyMediaClass extends \Spatie\MediaLibrary\Media
{

    public function getTypeAttribute()
    {
        $type = parent::getTypeAttribute();

        if ($type == self::TYPE_OTHER) {
            return $this->getTypeBasedOnMimeType();
        }
    }

    private function getTypeBasedOnMimeType()
    {
        // determine mime-type of the file and return one of
        // the self::TYPE_* constants based on the results
    }
}
```

**Comment by freekmurze** [comment]
Nice catch. You're right the type of file without an extension isn't detected properly. It would be better if the medialibrary should use `File::getMimeType()` first before relying on the extension to detect the type.

I'll try to look into this soon, but if you need this soon create a PR. It'll have a high probability of being pulled :smile:

**Comment by cviebrock** [comment]
I'll probably have a PR for you tomorrow. However, I was going to suggest checking the extension _first_, since it doesn't need to touch the file system to do so. If it can't figure it out by extension (or maybe, more appropriately, if it doesn't have an extension) then it would try to check the MIME type by actually looking at the file.

Make sense?

**Comment by freekmurze** [comment]
Makes perfect sense. Looking forward to that PR.

**Comment by freekmurze** [comment]
This is fixed in v3.17.1
